### PR TITLE
[shuffle] updated Network struct to support Trove Testnet

### PR DESCRIPTION
**Pull request by avin1100**
## Motivation

As we're ramping up to connect to Trove Testnet, I've refactored the Networks struct in shared.rs and added Trove Testnet information to the Networks.toml. 

## Test Plan

image

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/diem/diem/tree/main/developers.diem.com, and link to your PR here.)

## If targeting a release branch, please fill the below out as well

 * Justification and breaking nature (who does it affect? validators, full nodes, tooling, operators, AOS, etc.)
 * Comprehensive test results that demonstrate the fix working and not breaking existing workflows.
 * Why we must have it for V1 launch.
 * What workarounds and alternative we have if we do not push the PR.

### CODE REVIEW

File: shuffle/cli/src/shared.rs [hunk 1]
```diff
@@ -25,8 +25,15 @@ use url::Url;
...
 pub const MAIN_PKG_PATH: &str = "main";
 const NEW_KEY_FILE_CONTENT: &[u8] = include_bytes!("../new_account.key");
-pub const LOCALHOST_NETWORK_NAME: &str = "localhost";
-pub const LOCALHOST_NETWORK_BASE: &str = "http://127.0.0.1";
+
+pub const LOCALHOST_NAME: &str = "localhost";
+pub const LOCALHOST_JSON_RPC_URL: &str = "http://127.0.0.1:8080";
+pub const LOCALHOST_DEV_API_URL: &str = "http://127.0.0.1:8081";
+
+pub const TROVE_TESTNET_NETWORK_NAME: &str = "trove";
```
Comment (dimroc) [inline shuffle/cli/src/shared.rs:33]: why are these all `pub` public?

File: shuffle/cli/src/shared.rs [hunk 2]
```diff
@@ -25,8 +25,15 @@ use url::Url;
...
+pub const TROVE_TESTNET_JSON_RPC_URL: &str = "http://trove.aws.hlw3truzy4ls.com/";
+pub const TROVE_TESTNET_DEV_API_URL: &str = "https://api.trove.aws.hlw3truzy4ls.com/";
+pub const TROVE_TESTNET_FAUCET_URL: &str = "https://trove.aws.hlw3truzy4ls.com/mint";
```
Comment (dimroc) [inline shuffle/cli/src/shared.rs:36]: We want the base urls for these endpoints, because we will build the desired REST endpoint from that. ie:
`POST https://base.com/accounts`.
1. Remove trailing `/mint`
2. Validate that you only have a base url

File: shuffle/cli/src/shared.rs [hunk 3]
```diff
@@ -308,8 +315,18 @@ impl Home {
 
 pub fn normalized_network(home: &Home, network: Option<String>) -> Result<Url> {
     match network {
-        Some(input) => Ok(home.read_networks_toml()?.url_for(input.as_str())?),
-        None => Ok(home.read_networks_toml()?.url_for(LOCALHOST_NETWORK_NAME)?),
+        Some(input) => Ok(home
+            .read_networks_toml()?
+            .networks
+            .get(input.as_str())
```
Comment (dimroc) [inline shuffle/cli/src/shared.rs:321]: In another PR, are we creating a `get` that you can use directly on all_networks?

File: shuffle/cli/src/shared.rs [hunk 4]
```diff
@@ -321,43 +338,44 @@ pub struct NetworksConfig {
...
-    }
 }
 
 #[derive(Debug, Serialize, Deserialize, PartialEq)]
 pub struct Network {
     name: String,
-    base: String,
-    json_rpc_port: u16,
-    dev_api_port: u16,
+    json_rpc_url: String,
```
Comment (dimroc) [inline shuffle/cli/src/shared.rs:352]: Can we cut json_rpc yet?

**Review by dimroc — CHANGES_REQUESTED** [review]

File: shuffle/cli/src/shared.rs (hunk 1 above)
Comment (avin1100) [reply shuffle/cli/src/shared.rs:33]: Was in process of removing all

File: shuffle/cli/src/shared.rs (hunk 4 above)
Comment (avin1100) [reply shuffle/cli/src/shared.rs:352]: we need json_rpc_url for FaucetClient.

File: shuffle/cli/src/shared.rs (hunk 2 above)
Comment (avin1100) [reply shuffle/cli/src/shared.rs:36]: Was a bit confused here. Used the faucet_endpoint from https://fb.quip.com/UytvAaPdWsu4. Using new faucet endpoint for account creation found here: https://github.com/diem/diem/pull/9568/files#diff-841273dacd152ce5926f77cdf78ac63171ba37b969726f5d4bfb63f5bf6ca060R123

File: shuffle/cli/src/shared.rs (hunk 3 above)
Comment (avin1100) [reply shuffle/cli/src/shared.rs]: added in this PR

File: shuffle/cli/src/shared.rs [hunk 5]
```diff
@@ -27,10 +27,17 @@ use url::Url;
...
 const DIEM_ACCOUNT_TYPE: &str = "0x1::DiemAccount::DiemAccount";
 
+const LOCALHOST_NAME: &str = "localhost";
+const LOCALHOST_JSON_RPC_URL: &str = "http://127.0.0.1:8080";
+const LOCALHOST_DEV_API_URL: &str = "http://127.0.0.1:8081";
+
+const TROVE_TESTNET_NETWORK_NAME: &str = "trove";
+const TROVE_TESTNET_JSON_RPC_URL: &str = "http://trove.aws.hlw3truzy4ls.com";
+const TROVE_TESTNET_DEV_API_URL: &str = "https://api.trove.aws.hlw3truzy4ls.com";
+const TROVE_TESTNET_FAUCET_URL: &str = "https://trove.aws.hlw3truzy4ls.com/accounts";
```
Comment (sunmilee) [inline shuffle/cli/src/shared.rs:39]: Why are we hardcoding these here vs reading from a config/env?

Comment (avin1100) [reply shuffle/cli/src/shared.rs:39]: These hard codings are written into the config once, and after that we only read from the config to get the Network values

Comment (sunmilee) [reply shuffle/cli/src/shared.rs:39]: If that's the case, I'm a little confused why we don't just hardcode into the config and get rid of these?

File: shuffle/cli/src/shared.rs [hunk 6]
```diff
@@ -407,44 +413,61 @@ pub struct NetworksConfig {
 }
 
 impl NetworksConfig {
-    pub fn default() -> Self {
+    pub fn default() -> Result<Self> {
         let mut network_map = BTreeMap::new();
-        network_map.insert("localhost".to_string(), Network::default());
-        NetworksConfig {
+        network_map.insert("localhost".to_string(), Network::localhost()?);
+        network_map.insert("trove".to_string(), Network::trove_testnet()?);
```
Comment (dimroc) [inline shuffle/cli/src/shared.rs:419]: let's have the key match the call. There might be a trove production that we will want to call trove.

```rust
        network_map.insert("trove_testnet".to_string(), Network::trove_testnet()?);
```

File: shuffle/cli/src/shared.rs [hunk 7]
```diff
@@ -407,44 +413,61 @@ pub struct NetworksConfig {
...
-            name: String::from(LOCALHOST_NETWORK_NAME),
-            base: String::from(LOCALHOST_NETWORK_BASE),
-            json_rpc_port: 8080,
-            dev_api_port: 8081,
-        }
+    pub fn localhost() -> Result<Self> {
+        Ok(Network {
+            name: String::from(LOCALHOST_NAME),
+            json_rpc_url: Url::from_str("http://127.0.0.1:8080")?,
+            dev_api_url: Url::from_str("http://127.0.0.1:8081")?,
```
Comment (dimroc) [inline shuffle/cli/src/shared.rs:445]: As of https://github.com/diem/diem/pull/9692 the default port for the dev api is 8080 (same as json rpc).

File: shuffle/cli/src/shared.rs [hunk 8]
```diff
@@ -407,44 +413,61 @@ pub struct NetworksConfig {
...
+        })
+    }
+
+    pub fn trove_testnet() -> Result<Self> {
+        Ok(Network {
+            name: String::from("trove"),
+            json_rpc_url: Url::from_str("http://trove.aws.hlw3truzy4ls.com")?,
+            dev_api_url: Url::from_str("https://api.trove.aws.hlw3truzy4ls.com")?,
+            faucet_url: Some(Url::from_str(
+                "https://trove.aws.hlw3truzy4ls.com/accounts",
```
Comment (dimroc) [inline shuffle/cli/src/shared.rs:456]: We touched on this before offline. `faucet_base_url` != `faucet_accounts_url`.

In the future, we'll be doing more than just account creation through the faucet, for example we'll be doing [funding of accounts](https://github.com/diem/diem/blob/main/client/faucet/src/lib.rs#L201-L209). If you only have `faucet_accounts_url`, it's clumsier to build `faucet_funding_url`, while if you have a `faucet_base_url`, we can construct both accounts and funding urls with relative paths.

TL;DR: Remove the trailing `/accounts` from the url.

**Review by dimroc — CHANGES_REQUESTED** [review]

File: shuffle/cli/src/shared.rs (hunk 5 above)
Comment (dimroc) [reply shuffle/cli/src/shared.rs]: This is for config generation, not reading. On first run on `shuffle node`, we have an empty folder and we need a string literal to seed the first ~/.shuffle/Networks.toml with defaults, like localhost, so future runs can read from Networks.toml.

File: shuffle/cli/src/shared.rs [hunk 9]
```diff
@@ -407,44 +414,59 @@ pub struct NetworksConfig {
...
+        network_map.insert(TROVE_TESTNET_NAME.to_string(), Network::trove_testnet()?);
+        Ok(NetworksConfig {
             networks: network_map,
-        }
+        })
     }
 
-    pub fn url_for(&self, network_name: &str) -> Result<Url> {
-        let specified_network = self.networks.get(network_name).ok_or_else(|| {
+    pub fn get_network(&self, network_name: &str) -> Result<Network> {
```
Comment (dimroc) [inline shuffle/cli/src/shared.rs:426]: nit: What's wrong with just `get()`? What else would we get?

File: shuffle/cli/src/shared.rs [hunk 10]
```diff
@@ -407,44 +414,59 @@ pub struct NetworksConfig {
...
+            faucet_url: None,
+        })
+    }
+
+    pub fn trove_testnet() -> Result<Self> {
+        Ok(Network {
+            name: String::from(TROVE_TESTNET_NAME),
+            json_rpc_url: Url::from_str("http://trove.aws.hlw3truzy4ls.com")?,
+            dev_api_url: Url::from_str("https://api.trove.aws.hlw3truzy4ls.com")?,
+            faucet_url: Some(Url::from_str("https://trove.aws.hlw3truzy4ls.com")?),
```
Comment (dimroc) [inline shuffle/cli/src/shared.rs:456]: We know these to be valid URLs, so we shouldn't have to worry about returning an error at runtime.

`Network::localhost()` and `Network::trove_testnet()` shouldn't return a `Result`. Perhaps we get risky and just call unwrap.

@bmwill what would you recommend here?

File: shuffle/cli/src/shared.rs [hunk 11]
```diff
@@ -771,49 +793,58 @@ mod test {
...
+        network_map.insert("localhost".to_string(), get_test_localhost_network());
+        network_map.insert(
+            "trove_testnet".to_string(),
+            get_test_trove_testnet_network(),
+        );
         let networks_cfg = NetworksConfig {
             networks: network_map,
         };
-        assert_eq!(NetworksConfig::default(), networks_cfg);
+        assert_eq!(NetworksConfig::default().unwrap(), networks_cfg);
```
Comment (dimroc) [inline shuffle/cli/src/shared.rs:828]: Shouldn't need an unwrap() here imo. cc @bmwill

**Review by dimroc — CHANGES_REQUESTED** [review]

File: shuffle/cli/src/shared.rs (hunk 9 above)
Comment (avin1100) [reply shuffle/cli/src/shared.rs:426]: True

File: shuffle/cli/src/shared.rs (hunk 10 above)
Comment (bmwill) [reply shuffle/cli/src/shared.rs:456]: These are not public endpoints and as such should not be committed into the repo.

File: shuffle/cli/src/shared.rs (hunk 11 above)
Comment (bmwill) [reply shuffle/cli/src/shared.rs:828]: Yeah if you have a method named `default` you should be implementing the `Default` trait who's signature does not return an error.

File: shuffle/cli/src/shared.rs [hunk 12]
```diff
@@ -407,44 +414,59 @@ pub struct NetworksConfig {
...
 }
 
 #[derive(Debug, Serialize, Deserialize, PartialEq)]
 pub struct Network {
     name: String,
-    base: String,
-    json_rpc_port: u16,
-    dev_api_port: u16,
+    json_rpc_url: Url,
+    dev_api_url: Url,
```
Comment (xli) [inline shuffle/cli/src/shared.rs:437]: `dev_api_url` serves for json-rpc as well, so I think we should remove `json_rpc_url`, use `dev_api_url` for json-rpc calls.

Comment (dimroc) [reply shuffle/cli/src/shared.rs:437]: That might not always be the case, as shown [here](https://fb.quip.com/UytvAaPdWsu4). @avin1100 mentioned that the rust faucet client expects a json rpc url, so there's still some lingering dependencies on it.

File: shuffle/cli/src/shared.rs [hunk 13]
```diff
@@ -407,43 +407,52 @@ pub struct NetworksConfig {
...
+    faucet_url: Option<Url>,
 }
 
 impl Network {
-    pub fn default() -> Self {
+    pub fn from(network_ref: &Network) -> Result<Self> {
```
Comment (dimroc) [inline shuffle/cli/src/shared.rs:436]: This functionality resembles that of `clone()`. Some thoughts:
1. Let's rename this to copy and implement the Clone trait: https://doc.rust-lang.org/std/clone/trait.Clone.html
2. Do you even need this? Can you simply `derive` clone: https://doc.rust-lang.org/std/clone/trait.Clone.html#derivable

**Review by dimroc — CHANGES_REQUESTED** [review]

File: shuffle/cli/src/shared.rs (hunk 13 above)
Comment (avin1100) [reply shuffle/cli/src/shared.rs:436]: This isn't a clone though. I'm changing the type between a reference and a value.

Comment (avin1100) [reply shuffle/cli/src/shared.rs:436]: This is so cool! Using this.

**Review by dimroc — DISMISSED** [review]
